Create a Quickstart to run a Tekton task with Minikube

geriom opened the pull request:
# Changes

Introductory guide friendly for newcomers. The approach in this doc is to be
easier to follow by providing prescriptive instructions about the tools
to use. After installing the prerequisites, doing copy/paste of the code
samples should work.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/website/blob/master/CONTRIBUTING.md)
for more details._

jjasghar commented:
Taking a look today, give me a bit of time to read over this.

bobcatfish commented on content/en/docs/Getting Started/quickstart-kind.md:
```diff
@@ -0,0 +1,168 @@
+
+## Create and run a basic Task
+
+A **Taks** defines a series of **steps** that run sequentially to perform some
```
typo taks -> task

geriom replied:
Fixed

bobcatfish commented on content/en/docs/Getting Started/quickstart-kind.md:
```diff
@@ -0,0 +1,168 @@
+	 			command:
+	 				- echo
+	 			args:
+	 				- "Hello World"
```
there's something odd about the indentation here - i think it's way to deep (2 spaces would probably be good, or 4 at most?) and it shows up as red in the preview 🤔 

![image](https://user-images.githubusercontent.com/432502/134740916-592f8f8c-444c-4b79-9a8d-eda2860ff873.png)

geriom replied:
I think these were tabs instead of spaces (?). This looked fine on my computer :). I made sure they were converted to spaces, I think it's fixed now.

bobcatfish commented on content/en/docs/Getting Started/quickstart-kind.md:
````diff
@@ -0,0 +1,168 @@
+   spec:
+     taskRef:
+       name: echo-hello-world
+   ```
````
probably makes sense to use TaskRuns explicitly but just so you know there's one other option that might be a nice user experience - but the user would have to also install `tkn` (https://github.com/tektoncd/cli) so maybe it's not worth the extra setup - but if it wasn't too much trouble, instead of making a TaskRun here the user could do [`tkn task start echo-hello-world`](https://github.com/tektoncd/cli/blob/main/docs/cmd/tkn_task_start.md)

geriom replied:
I thought about it, same to check the task status. The trade-off of using a simpler command but having to install something extra, if this is your first ever experience with tekton, may not be worth it in my opinion. For the longer tutorial covering pipelines in more detail it's definitely a good idea to ask for the `tkn` client prerequisite.

bobcatfish commented on content/en/docs/Getting Started/quickstart-kind.md:
```diff
@@ -0,0 +1,168 @@
+     ...
+     </pre>
+
+     The line `Status: Succeeded` confirms that TaskRun completed with no errors.
```
instead of showing them the pod, it might be a bit nicer to show them the taskrun, e.g. `kubectl --context kind-tektoncd get taskrun echo-hello-world-task-run` would show something simple like:

```
NAME                               SUCCEEDED    REASON       STARTTIME   COMPLETIONTIME
echo-hello-world-task-run          True         Succeeded    22h         22h
```

plus the pods are (kind of) an implementation detail of tekton anyway 🤞  (tho in reality is not like ppl can really ignore them...)

geriom replied:
Updated

jjasghar commented:
My only question, is this to show off Tekton _inside_ of KiND? Otherwise I'm not seeing the advantage of this over https://tekton.dev/docs/getting-started/pipelines/

bobcatfish reviewed (commented):
Very cool!! I was able to run through it myself so easily (i might be not quite the target audience tho haha - BUT ive never really used KIND before so that was pretty cool for me!)

Couple of thoughts:
* This is kind of a nit but I'm wondering if we want to have this repo as the source of truth for this quickstart - it seems like it's only using Tekton Pipelines, would it make sense to store it (and potentially version and update it) alongside the Tekton Pipelines docs? (maybe it could even replace [the tutorial over there?)](https://github.com/tektoncd/pipeline/blob/main/docs/tutorial.md)
* What do you think about updating or maybe replacing the existing getting started landing page? (https://tekton.dev/docs/getting-started/) (seems like this kind tutorial would be a good replacement?)

bobcatfish commented:
@jjasghar i think what's really nice about using kind is that it's very easy to setup and lets us have getting started instructions that are pretty cloud agnostic (but still tell you how to setup the cluster in the first place!)

my understanding is that this tutorial _would_ replace at least 1 (if not more - we have some duplication going on) of our existing tutorials. the one you linked to (https://tekton.dev/docs/getting-started/pipelines/) i think is meant to build on a previous task based tutorial and add pipelines? (presumably it would work on the kind cluster as well? maybe needs a few updates to flow nicely from the kind tutorial).  i think this tutorial would probably replace https://tekton.dev/docs/getting-started/ (maybe we go back to our old approach and have a tab or something for the openshift specific instructions?)

geriom commented:
@jjasghar This is intended to be the very first Tekton doc for someone with no prior experience. A 5-minute task that is easy to complete so you get that "feels good" moment that you need to explore further. That's why it's prescriptive about Kind, it's an effort to guarantee the task can be completed successfully with minimal effort. The existing getting started requires some cognitive work to follow.

@bobcatfish You are correct, my intention is to consolidate the doc in the pipelines repo (https://github.com/tektoncd/pipeline/blob/main/docs/tutorial.md), the 2 getting started docs from tekton.dev (getting-started and getting-started/pipelines), and the Katacoda tuto (https://github.com/tektoncd/website/tree/main/tutorials/katacoda/getting-started) in to a single doc that we can link from everywhere. The quickstart in his PR being the entry point and a further more in-depth doc as the natural follow-up.


I'm going to take a look at the other comments and apply the fixes later this week. Thanks for the feedback.

afrittoli commented:
Thanks @geriom - I haven't reviewed the instructions in details yet - but I definitely support using `kind` - that's what I use for my development work and POCs anyways. We have a script for that so perhaps we could include a link to it? https://github.com/tektoncd/plumbing/blob/main/hack/tekton_in_kind.sh

geriom commented:
> * This is kind of a nit but I'm wondering if we want to have this repo as the source of truth for this quickstart - it seems like it's only using Tekton Pipelines, would it make sense to store it (and potentially version and update it) alongside the Tekton Pipelines docs? (maybe it could even replace [the tutorial over there?)](https://github.com/tektoncd/pipeline/blob/main/docs/tutorial.md)

If we move it to that repo it would be harder to make it a landing page because of how the sync script works.

> * What do you think about updating or maybe replacing the existing getting started landing page? (https://tekton.dev/docs/getting-started/) (seems like this kind tutorial would be a good replacement?)

I like the idea. Should I proceed and move the current content to a different file?

bobcatfish commented:
> If we move it to that repo it would be harder to make it a landing page because of how the sync script works.

😨  well maybe in the meantime we could link directly to the page on tekton.dev from the pipelines repo :S

> Should I proceed and move the current content to a different file?

I'd remove the existing file entirely - in the end I think our docs at tekton.dev should be pretty authorative vs. kind of a grab bag of different tutorials. The tutorial will also still be in version control if someone wants to resurrect it later.

bobcatfish commented:
... did not mean to hit that button

bobcatfish commented:
TOO MUCH POWER

afrittoli commented:
/hold

afrittoli commented:
Waiting for the doc plan to be agreed on, then we can resume this one and get it merged

geriom commented:
@afrittoli This PR is ready to be merged after a a second round of reviews.

Updates:

I switched from Kind to Minikube because of the recent changes in the Docker licence. With Minikube the users can pick a different engine and are not forced to use docker to follow the instructions.

I also added a second page, this is now a two-part tutorial, somewhat mirroring the content we had before:

- Create a task
- Create a second task and a pipeline.

I also deleted the original tutorial that we had in `_index` and made it a simple landing page.

AlanGreene commented:
Looks like this removes reference to the nightly releases. It would be good to restore that as we do want to make them more visible. Ref: https://github.com/tektoncd/website/pull/344#issuecomment-1038992697

geriom commented:
In the spirit of the [new guidelines for tutorials](https://tekton.dev/docs/contribute/doc-con-content/), I think we should pick just one path and go with it throughout the entire doc, don't cover alternatives in the getting started. Instead, I would prefer to create an installation doc, that covers all the available installation options.

bobcatfish commented on content/en/docs/Getting Started/tasks.md:
```diff
@@ -0,0 +1,168 @@
+## Create and run a basic task
+
+A **task**, represented in the API as an object of kind `Task`, defines a
+series of **steps** that run sequentially to perform some build work. Every
```
nit: maybe instead of "build work" something more generic (but CI/CD specific), something like "perform logic the Task requires" or drop that part of the sentence entirely? (cuz not every Task is doing something build related)

bobcatfish commented on content/en/docs/Getting Started/tasks.md:
```diff
@@ -0,0 +1,168 @@
+	 hello-task-run          True         Succeeded    22h         22h
+     </pre>
+
+     The value `True` under `SUCCEEDED` confirms that TaskRun completed with no errors.
```
nit: might be nice to also look at the logs at this point, something like `k logs --selector=tekton.dev/taskRun=hello-task-run` but np if that's just complicating things!

geriom replied:
Updated

bobcatfish commented on content/en/docs/Getting Started/pipelines.md:
```diff
@@ -1,191 +1,165 @@
+	          image: ubuntu
+	          script: |
+	             #!/usr/bin/env bash
+	             echo "Goodbye World!"
```
there's some extra whitespace at the beginning of the yaml block that shows up when i try to copy paste from this block:

![image](https://user-images.githubusercontent.com/432502/154171090-7b508ac6-fc8e-43ba-854e-001994db31da.png)

bobcatfish commented on content/en/docs/Getting Started/pipelines.md:
```diff
@@ -1,191 +1,165 @@
+          - name: goodbye
+            runAfter:
+               - hello
+            TaskRef:
```
should be `taskRef`

(i think this will work initially but it leads to really weird bugs X'D https://github.com/tektoncd/triggers/issues/526)

geriom replied:
Fixed!

bobcatfish reviewed (commented):
very nice!!! just some minor comments from me, that was very smooth for me (and the first time ive used minikube in... years XD might do it more often now!)

(i'm going to hold off on doing an approve or lgtm and defer to folks that have been attending the docs WG but lemme know if I need to)

afrittoli reviewed (commented):
Thanks for this - I added some cosmetic comments that needs to be fixed before this is merged, but the content looks great.

Historically with always been starting our Tekton guides with writing your own `Task` and `Pipeline`... I wonder however if we could rethink that as an "advanced" step - and start with using `Tasks` from the catalog instead? Not something for this PR, but I wanted to bring up the idea. @tektoncd/core-maintainers thoughts?

bobcatfish commented:
> Historically with always been starting our Tekton guides with writing your own Task and Pipeline... I wonder however if we could rethink that as an "advanced" step - and start with using Tasks from the catalog instead? Not something for this PR, but I wanted to bring up the idea. https://github.com/orgs/tektoncd/teams/core-maintainers thoughts?

that could make a lot of sense @afrittoli  - could get users much more quickly to the satisfaction of running something and having it work :D

geriom commented on content/en/docs/Getting Started/tasks.md:
````diff
@@ -0,0 +1,180 @@
+           - echo
+         args:
+           - "Hello World"
+   ```
````
I updated the sample to use `script` instead.,

geriom commented on content/en/docs/Getting Started/pipelines.md:
````diff
@@ -1,191 +1,165 @@
+	         script: |
+	            #!/usr/bin/env bash
+	            echo "Goodbye World!"
+   ```
````
Somehow I ended up with tabs instead of spaced. It should be fixed now.

geriom commented on content/en/docs/Getting Started/tasks.md:
````diff
@@ -0,0 +1,180 @@
+
+     The output confirms that the task was completed successfully.
+
+     ```bash
+     task.tekton.dev/hello created
+     ```
````
All the rendering glitches can be fixed using a newer version of hugo. #361 Fixes this.

geriom commented on content/en/docs/Getting Started/tasks.md:
````diff
@@ -0,0 +1,180 @@
+   https://storage.googleapis.com/tekton-releases/pipeline/latest/release.yaml
+   ```
+
+1. Monitor the installation:
````
Fixed by #361

geriom commented:
Completely agree with both of you @bobcatfish and @afrittoli . I'm going to give it some thought and share my ideas in either the WG or the mailing list.

afrittoli commented:
@geriom @AlanGreene I kicked off a re-deploy now, but it did not help - I still see the same cosmetic glitches in the rendering. I think this needs to be rebased so we can preview the actual result.

Apart from that the content looks good to me, so feel free to merge once the cosmetic issues are fixed

geriom commented:
@afrittoli I rebased the branch, the preview looks fine now.

@AlanGreene I think we can merge it.

AlanGreene commented on content/en/docs/Getting Started/tasks.md:
````diff
@@ -0,0 +1,178 @@
+1. To run this task, you must instantiate it using `TaskRun`. Create another
+   file named `hello-world-run.yaml` with the following content:
+
+   ```yaml
````
similar indentation issue with this block when rendered

geriom replied:
Fixed

geriom commented on content/en/docs/Getting Started/tasks.md:
```diff
@@ -0,0 +1,178 @@
+
+1.  [Install minikube](https://minikube.sigs.k8s.io/docs/start/). You only have
+    to complete the step 1, "Installation".
+1.  [Install kubectl](https://minikube.sigs.k8s.io/docs/start/)
```
Good catch! That's is the right link, thank you.

geriom replied:
Good catch! That is indeed the correct link

geriom commented on content/en/docs/Getting Started/tasks.md:
````diff
@@ -0,0 +1,178 @@
+To delete the cluster that you created for this quickstart run:
+
+```bash
+kind delete cluster --name tektoncd
````
Fixed

AlanGreene reviewed (approved):
Looks good, thanks @geriom 

/lgtm
/hold cancel
/meow
